File: backend/utils/source_titles.py
"""Format vector source citations for storage, API telemetry, and UI display."""

from __future__ import annotations

import re
from typing import Any, Dict, List, Optional, Tuple

DISPLAY_TITLE_MAX_LEN = 48

TRAILING_PUNCTUATION_RE = re.compile(r"[.;:\s]+$")
STORED_ELLIPSIS_RE = re.compile(r"\.{2,}")


def normalize_stored_citation_field(value: str) -> str:
    """Normalize one stored citation field: no trailing punctuation or ellipsis."""
    cleaned = str(value or "").strip()
    if not cleaned:
        return ""

    cleaned = STORED_ELLIPSIS_RE.sub("", cleaned).strip()
    return TRAILING_PUNCTUATION_RE.sub("", cleaned).strip()


def compose_source_citation(
    title: str,
    *,
    volume: Optional[str] = None,
    chapter: Optional[str] = None,
) -> str:
    """Join stored citation parts with the canonical ` - ` separator."""
    parts: List[str] = []
    normalized_title = normalize_stored_citation_field(title)
    if normalized_title:
        parts.append(normalized_title)

    normalized_volume = normalize_stored_citation_field(volume or "")
    if normalized_volume:
        parts.append(normalized_volume)

    normalized_chapter = normalize_stored_citation_field(chapter or "")
    if normalized_chapter:
        parts.append(normalized_chapter)

    return " - ".join(parts)


def format_source_title(title: str, *, max_len: int = DISPLAY_TITLE_MAX_LEN) -> str:
    """Strip trailing punctuation and truncate long composed citation strings for display."""
    cleaned = title.strip()
    if not cleaned:
        return ""

    # Backward compatibility for legacy vector rows that still store trailing ellipsis.
    if not cleaned.endswith("..."):
        cleaned = TRAILING_PUNCTUATION_RE.sub("", cleaned).strip()

    if len(cleaned) <= max_len:
        return cleaned

    if max_len <= 3:
        return "..."

    return cleaned[: max_len - 3].rstrip() + "..."


def format_source_citation(
    title: str,
    *,
    volume: Optional[str] = None,
    chapter: Optional[str] = None,
    max_len: int = DISPLAY_TITLE_MAX_LEN,
) -> str:
    """Compose a citation and truncate the title while keeping volume/chapter visible."""
    normalized_title = normalize_stored_citation_field(title)
    suffix_parts: List[str] = []

    normalized_volume = normalize_stored_citation_field(volume or "")
    if normalized_volume:
        suffix_parts.append(normalized_volume)

    normalized_chapter = normalize_stored_citation_field(chapter or "")
    if normalized_chapter:
        suffix_parts.append(normalized_chapter)

    if not suffix_parts:
        return format_source_title(normalized_title, max_len=max_len)

    suffix = " - " + " - ".join(suffix_parts)
    if not normalized_title:
        return format_source_title(suffix.removeprefix(" - "), max_len=max_len)

    composed = normalized_title + suffix
    if len(composed) <= max_len:
        return composed

    title_budget = max_len - len(suffix)
    if title_budget <= 3:
        return format_source_title(composed, max_len=max_len)

    return format_source_title(normalized_title, max_len=title_budget) + suffix


def citation_from_metadata(metadata: Dict[str, Any]) -> Dict[str, Optional[str]]:
    """Build a structured citation dict from one vector chunk metadata row."""
    title = normalize_stored_citation_field(str(metadata.get("source_title") or ""))
    volume = normalize_stored_citation_field(str(metadata.get("source_volume") or ""))
    chapter = normalize_stored_citation_field(str(metadata.get("source_chapter") or ""))
    return {
        "title": title,
        "volume": volume or None,
        "chapter": chapter or None,
    }


def citation_dedupe_key(citation: Dict[str, Optional[str]]) -> Tuple[str, str, str]:
    """Stable dedupe key for telemetry source lists."""
    return (
        citation.get("title") or "",
        citation.get("volume") or "",
        citation.get("chapter") or "",
    )

File: frontend/components/source-titles.ts
// Formats vector source labels for bubble footers. Mirrors backend/utils/source_titles.py.

const DISPLAY_TITLE_MAX_LEN = 48;
const TRAILING_PUNCTUATION_RE = /[.;:\s]+$/;
const STORED_ELLIPSIS_RE = /\.{2,}/;

export type VectorSourceCitation = {
  title: string;
  volume?: string | null;
  chapter?: string | null;
};

export function normalizeStoredCitationField(value: string): string {
  let cleaned = value.trim();
  if (!cleaned) {
    return "";
  }

  cleaned = cleaned.replace(STORED_ELLIPSIS_RE, "").trim();
  return cleaned.replace(TRAILING_PUNCTUATION_RE, "").trim();
}

export function composeSourceCitation(
  title: string,
  volume?: string | null,
  chapter?: string | null,
): string {
  const parts: string[] = [];
  const normalizedTitle = normalizeStoredCitationField(title);
  if (normalizedTitle) {
    parts.push(normalizedTitle);
  }

  const normalizedVolume = normalizeStoredCitationField(volume ?? "");
  if (normalizedVolume) {
    parts.push(normalizedVolume);
  }

  const normalizedChapter = normalizeStoredCitationField(chapter ?? "");
  if (normalizedChapter) {
    parts.push(normalizedChapter);
  }

  return parts.join(" - ");
}

export function formatSourceTitle(title: string, maxLen = DISPLAY_TITLE_MAX_LEN): string {
  let cleaned = title.trim();
  if (!cleaned) {
    return "";
  }

  if (!cleaned.endsWith("...")) {
    cleaned = cleaned.replace(TRAILING_PUNCTUATION_RE, "").trim();
  }

  if (cleaned.length <= maxLen) {
    return cleaned;
  }

  if (maxLen <= 3) {
    return "...";
  }

  return `${cleaned.slice(0, maxLen - 3).trimEnd()}...`;
}

export function formatSourceCitation(
  citation: VectorSourceCitation | string,
  maxLen = DISPLAY_TITLE_MAX_LEN,
): string {
  if (typeof citation === "string") {
    return formatSourceTitle(citation, maxLen);
  }

  const normalizedTitle = normalizeStoredCitationField(citation.title);
  const suffixParts: string[] = [];

  const normalizedVolume = normalizeStoredCitationField(citation.volume ?? "");
  if (normalizedVolume) {
    suffixParts.push(normalizedVolume);
  }

  const normalizedChapter = normalizeStoredCitationField(citation.chapter ?? "");
  if (normalizedChapter) {
    suffixParts.push(normalizedChapter);
  }

  if (suffixParts.length === 0) {
    return formatSourceTitle(normalizedTitle, maxLen);
  }

  const suffix = ` - ${suffixParts.join(" - ")}`;
  if (!normalizedTitle) {
    return formatSourceTitle(suffix.slice(3), maxLen);
  }

  const composed = normalizedTitle + suffix;
  if (composed.length <= maxLen) {
    return composed;
  }

  const titleBudget = maxLen - suffix.length;
  if (titleBudget <= 3) {
    return formatSourceTitle(composed, maxLen);
  }

  return formatSourceTitle(normalizedTitle, titleBudget) + suffix;
}

function citationKey(citation: VectorSourceCitation | string): string {
  if (typeof citation === "string") {
    return formatSourceTitle(citation);
  }

  return [
    normalizeStoredCitationField(citation.title),
    normalizeStoredCitationField(citation.volume ?? ""),
    normalizeStoredCitationField(citation.chapter ?? ""),
  ].join("\0");
}

export function normalizeSourceTitles(
  sources: Array<string | VectorSourceCitation> | undefined | null,
): string[] {
  const seen = new Set<string>();
  const normalized: string[] = [];

  for (const source of sources ?? []) {
    const formatted = formatSourceCitation(source);
    const key = citationKey(source);
    if (!formatted || seen.has(key)) {
      continue;
    }

    seen.add(key);
    normalized.push(formatted);
  }

  return normalized;
}
